File: project-java-cosminpac04/motorcycleContest/src/main/java/motorcycle/repository/RaceRepository.java
package motorcycle.repository;

import motorcycle.database.DataBaseConnection;
import motorcycle.model.Participant;
import motorcycle.model.Race;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class RaceRepository implements IRepository<Race, Integer> {
    private static final Logger logger = LoggerFactory.getLogger(RaceRepository.class);

    @Override
    public void add(Race race) {
        String sql = "INSERT INTO races (Capacity) VALUES (?)";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {

            stmt.setInt(1, race.getCapacity());
            stmt.executeUpdate();

            // Retrieve generated ID
            ResultSet rs = stmt.getGeneratedKeys();
            if (rs.next()) {
                race.setID(rs.getInt(1));
            }

            logger.info("Added race with capacity: {}", race.getCapacity());

        } catch (SQLException e) {
            logger.error("Error adding race", e);
        }
    }

    @Override
    public void delete(Race race) {
        String sql = "DELETE FROM races WHERE Rid = ?";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, race.getID());
            stmt.executeUpdate();

            logger.info("Deleted race with ID: {}", race.getID());

        } catch (SQLException e) {
            logger.error("Error deleting race with ID: {}", race.getID(), e);
        }
    }

    @Override
    public void update(Race race, Integer id) {
        String sql = "UPDATE races SET Capacity = ? WHERE Rid = ?";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, race.getCapacity());
            stmt.setInt(2, id);
            stmt.executeUpdate();

            logger.info("Updated race with ID: {}", id);

        } catch (SQLException e) {
            logger.error("Error updating race with ID: {}", id, e);
        }
    }

    @Override
    public Race findById(Integer id) {
        String sql = "SELECT * FROM races WHERE Rid = ?";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, id);
            ResultSet rs = stmt.executeQuery();

            if (rs.next()) {
                Race race = new Race(rs.getInt("id"), rs.getInt("capacity"));
                race.setParticipants(getParticipantsForRace(id));
                logger.info("Found race with ID: {}", id);
                return race;
            } else {
                logger.warn("No race found with ID: {}", id);
            }

        } catch (SQLException e) {
            logger.error("Error finding race with ID: {}", id, e);
        }
        return null;
    }

    @Override
    public List<Race> findAll() {
        List<Race> races = new ArrayList<>();
        String sql = "SELECT * FROM races";

        try (Connection conn = (Connection) DataBaseConnection.connect();
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(sql)) {

            while (rs.next()) {
                Race race = new Race(rs.getInt("id"), rs.getInt("capacity"));
                race.setParticipants(getParticipantsForRace(race.getID()));
                races.add(race);
            }

            logger.info("Retrieved {} races", races.size());

        } catch (SQLException e) {
            logger.error("Error retrieving races", e);
        }
        return races;
    }

    @Override
    public List<Race> getAll() {
        return findAll();
    }

    // Helper method to fetch participants for a given race
    private List<Participant> getParticipantsForRace(int raceId) {
        List<Participant> participants = new ArrayList<>();
        String sql = "SELECT p.Pid, p.Name, p.EngineCapacity, p.Team " +
                "FROM participants p " +
                "JOIN race_participants rp ON p.Pid = rp.participant_id " +
                "WHERE rp.race_id = ?";

        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, raceId);
            ResultSet rs = stmt.executeQuery();

            while (rs.next()) {
                participants.add(new Participant(
                        rs.getInt("id"),
                        rs.getString("name"),
                        rs.getInt("engine_capacity"),
                        rs.getString("team")
                ));
            }

            logger.info("Retrieved {} participants for race ID {}", participants.size(), raceId);

        } catch (SQLException e) {
            logger.error("Error retrieving participants for race ID {}", raceId, e);
        }
        return participants;
    }
}

File: project-c-cosminpac04/motorcycleApp/Repositories/RaceRepository.cs
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using motorcycleApp.DataBase;
using motorcycleApp.Models;
using Serilog;

namespace motorcycleApp.Repositories;

public class RaceRepository : IRepository<Race, int> {
    public void Add(Race entity) {
        string sql = "INSERT INTO Race (Capacity) VALUES (@Capacity)";
        try {
            using (var conn = ConnectionManager.GetConnection()) {
                Log.Information("Connection established " + conn.Database);
                
                using (var command = new SQLiteCommand(sql, conn)) {
                    command.Parameters.AddWithValue("@Capacity", entity.Capacity);
                    command.ExecuteNonQuery();
                    Log.Information("Added race with capacity " + entity.Capacity + " to Database");
                }
            }
        }
        catch (Exception ex) {
            Log.Error(ex, "Error while adding race with capacity " + entity.Capacity + " to Database");
        }
    }

    public void Update(Race entity) {
        string sql = "UPDATE Race SET Capacity=@Capacity WHERE ID=@Id";

        using (var conn = ConnectionManager.GetConnection()) {
            Log.Information("Connection established " + conn.Database);
            
            using (var command = new SQLiteCommand(sql, conn)) {
                command.Parameters.AddWithValue("@Id", entity.ID);
                command.Parameters.AddWithValue("@Capacity", entity.Capacity);
                command.ExecuteNonQuery();
                Log.Information("Updated race with ID " + entity.ID + " to capacity " + entity.Capacity);
            }
        }
    }

    public void Delete(Race entity) {
        string sql = "DELETE FROM Race WHERE ID=@Id";

        using (var conn = ConnectionManager.GetConnection()) {
            Log.Information("Connection established " + conn.Database);

            using (var command = new SQLiteCommand(sql, conn)) {
                command.Parameters.AddWithValue("@Id", entity.ID);
                int affectedRows = command.ExecuteNonQuery();
                
                if (affectedRows > 0) {
                    Log.Information("Deleted race with ID " + entity.ID + " from Database");
                }
                else {
                    Log.Warning("Attempted to delete race with ID " + entity.ID + " but no rows were affected");
                }
            }
        }
    }

    public Race FindById(int id) {
        string sql = "SELECT * FROM Race WHERE ID=@Id";

        using (var conn = ConnectionManager.GetConnection()) {
            Log.Information("Connection established " + conn.Database);

            using (var command = new SQLiteCommand(sql, conn)) {
                command.Parameters.AddWithValue("@Id", id);
                using (var reader = command.ExecuteReader()) {
                    if (reader.Read()) {
                        var race = new Race(
                            reader.GetInt32(0), // ID
                            reader.GetInt32(1)  // Capacity
                        );
                        Log.Information("Found race with ID " + id);
                        return race;
                    }
                }
            }
        }
        Log.Warning("Could not find race with ID " + id + " in Database");
        return null;
    }

    public IEnumerable<Race> GetAll() {
        string sql = "SELECT * FROM Race";
        var races = new List<Race>();

        using (var conn = ConnectionManager.GetConnection()) {
            Log.Information("Connection established " + conn.Database);

            using (var command = new SQLiteCommand(sql, conn))
            using (var reader = command.ExecuteReader()) {
                while (reader.Read()) {
                    races.Add(new Race(
                        reader.GetInt32(0), // ID
                        reader.GetInt32(1)  // Capacity
                    ));
                }
            }
        }
        Log.Information("Found " + races.Count + " races in Database");
        return races;
    }
}
